source/portfolio/portfolio_mad.cpp

```diff
@@ -0,0 +1,58 @@
+//
+// Created by eduardo on 15/04/2021.
+//
+
+#include "portfolio_mad.h"
+#include "portfolio/common/algorithm.h"
+#include <iostream>
+#include <numeric>
+#include <random>
+#include <range/v3/core.hpp>
+#include <range/v3/numeric/accumulate.hpp>
+namespace portfolio {
+    portfolio_mad::portfolio_mad(const market_data &data,
+                                 interval_points interval, int n_periods)
+        : interval_(std::move(interval)) {
+        n_periods_ = n_periods;
+        for (auto a = data.assets_map_begin(); a != data.assets_map_end();
```

Comment:
`for (const auto& [k,v]:  data)` não funciona?

Você pode definir um begin() e end() para suas próprias classes.